### PR TITLE
Create course project

### PR DESCRIPTION
**Pull request by nikke234**
# Description
Adds a "Turn Project Into A+ Project" action which does the following:
  + Select a course
  + Read the course configuration file
  + Import project settings
  + Import IDE settings (if they haven't already been imported for the course)
  + Add a "special" file to the project, such that the plugin recognizes which course the project belongs to when it is opened again

# Testing

- [ ] I created new unit tests
- [x] All unit tests pass
- [ ] I have tested manually

# Have you updated the README or other relevant documentation?

- [ ] Yes
- [ ] Not yet. I will do it next.
- [x] Not relevant

### CODE REVIEW

**Comment by superseacat** [comment]
I guess this will also need some changes in the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/create-course-project/TESTING.md)?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+    // Importing IDE settings potentially restarts the IDE, so it's the last action. If the
+    // IDE settings for the course have already been imported, do nothing.
+    if (!PluginSettings.getInstance().getImportedIdeSettingsName().equals(course.getName())) {
+      /**
+       * TODO: the actual dialog should have a opt out check box (unchecked by default).
+       */
+      dialogs.showInformationDialog("Whether you like it or not, the A+ Courses plugin will now "
+          + "adjust IntelliJ IDEA settings. This helps use IDEA for coursework. You can not opt "
+          + "out even if you want to.", "Adjust IDEA Settings");
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:84]: this is a good one :smile:

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  /**
+   * Creates a file in the project settings directory which contains the given course configuration
+   * file URL.
+   * @return True if the file was successfully created, false otherwise.
+   */
+  private boolean tryCreateCourseFile(@NotNull Project project, @NotNull URL courseUrl) {
+    try {
+      File file = new APlusProject(project).getCourseFilePath().toFile();
+      FileUtils.writeStringToFile(file, courseUrl.toString(), StandardCharsets.UTF_8);
+      return true;
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:147]: this has little to do with the implementation itself, but some points to consider about the file created:
* shall it have some known extension
* shall it be of some structured type (XML, JSON, etc.)

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+      return null;
+    }
+  }
+
+  /**
+   * Creates a file in the project settings directory which contains the given course configuration
+   * file URL.
+   * @return True if the file was successfully created, false otherwise.
+   */
+  private boolean tryCreateCourseFile(@NotNull Project project, @NotNull URL courseUrl) {
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:143]: the current (only) contents of the course are coming hard-coded with the plugin, right? why exactly do we need this file?

File: src/main/java/fi/aalto/cs/apluscourses/ui/repl/ReplConfigurationForm.java [hunk 4]
```diff
@@ -43,7 +43,8 @@ public ReplConfigurationForm() {
    */
   public ReplConfigurationForm(@NotNull ReplConfigurationFormModel model) {
     this.model = model;
-    dontShowThisWindowCheckBox.setSelected(PluginSettings.isShowReplConfigurationDialog());
+    dontShowThisWindowCheckBox.setSelected(
+        PluginSettings.getInstance().shouldShowReplConfigurationDialog());
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/ui/repl/ReplConfigurationForm.java:47]: why the change? :confused:

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        }
+      }
+    }
+  }
+
+  @Override
+  public void update(@NotNull AnActionEvent e) {
+    // This action is available only if a non-default project is open
+    Project project = e.getProject();
+    e.getPresentation().setEnabledAndVisible(project != null && !project.isDefault());
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:99]: what is this 'default' project?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  @Nullable
+  private URL getSelectedCourseUrl() {
+    // TODO: show a dialog with a list of courses and a URL field for custom courses, from which
+    // the user selects a course.
+    try {
+      URL selectedCourseUrl = new URL(PluginSettings.COURSE_CONFIGURATION_FILE_URL);
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:107]: IDE suggest some optimization here

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+      return selectedCourseUrl;
+    } catch (MalformedURLException e) {
+      // User entered an invalid URL (or the default list has an invalid URL, which would be a bug)
+      return null;
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:111]: at least `ex` and the stacktrace to the `log.err` :wink:

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+   * Construct a course project action with the given main view model provider and dialogs.
+   */
+  public CourseProjectAction(@NotNull MainViewModelProvider mainViewModelProvider,
+                             @NotNull Dialogs dialogs) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.dialogs = dialogs;
+  }
+
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent e) {
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:55]: Nit: it might be more convenient e.g. for testing purposes to break this longer method into a smaller chunks

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java (hunk 2 above)
Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:147]: I agree :+1: For now since the file only stores one URL, I think it's fine as is, but if more information is added then some file format is the way to go

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java (hunk 3 above)
Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:143]: The issue #86 includes a dialog from which the course is selected (with a URL field for unlisted courses). This file is needed to support multiple courses, as the plugin needs to know which project belongs to which course configuration.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java (hunk 5 above)
Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:99]: I'm not 100% sure either :smile: But `Project#getBasePath` [returns null for the default project](https://upsource.jetbrains.com/idea-ce/file/idea-ce-9276f4bbfb20dac51cdc0d675aef80b7f639a378/platform/core-api/src/com/intellij/openapi/project/Project.java), so `APlusProject#getCourseFilePath` would cause a null-pointer exception

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java [hunk 9]
```diff
@@ -25,6 +25,11 @@ public Path getBasePath() {
     return Paths.get(Objects.requireNonNull(project.getBasePath()));
   }
 
+  @NotNull
+  public Path getCourseFilePath() {
+    return getBasePath().resolve(Paths.get(Project.DIRECTORY_STORE_FOLDER, "a-plus-project"));
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java:30]: `FileUtilRt.toSystemIndependentName(...)` or will work wo?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java [hunk 10]
```diff
@@ -25,6 +25,11 @@ public Path getBasePath() {
     return Paths.get(Objects.requireNonNull(project.getBasePath()));
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java:25]: `FileUtilRt.toSystemIndependentName(...)` or will work wo?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/utils/SettingsUtil.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+    Path settingsPath = Paths.get(project.getBasePath(), Project.DIRECTORY_STORE_FOLDER);
+
+    File settingsZip = FileUtilRt.createTempFile(project.getName() + "-settings", ".zip");
+    CoursesClient.fetchZip(settingsUrl, settingsZip);
+    ZipFile zipFile = new ZipFile(settingsZip);
+    List<String> fileNames = zipFile
+        .getFileHeaders()
+        .stream()
+        .filter(file -> !file.isDirectory())
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/utils/SettingsUtil.java:81]: should here be some extra filtering step for checking the `fileHeaders` against `null` or something like that?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/utils/SettingsUtil.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+   *                                     course configuration.
+   */
+  public static void importProjectSettings(@NotNull Project project, @NotNull Course course)
+      throws IOException, UnexpectedResponseException {
+    URL settingsUrl = course.getResourceUrls().get("projectSettings");
+    if (settingsUrl == null) {
+      return;
+    }
+
+    Path settingsPath = Paths.get(project.getBasePath(), Project.DIRECTORY_STORE_FOLDER);
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/utils/SettingsUtil.java:73]: `project.getBasePath()` can be `null`...

Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/utils/SettingsUtil.java:73]: For the default project yes, not sure about other cases

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java (hunk 9 above)
Comment (OlliKiljunen) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java:30]: Java's `Path` class handles paths in a system independent way.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java (hunk 10 above)
Comment (OlliKiljunen) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java:25]: Java's Path class handles paths in a system independent way.

File: src/main/java/fi/aalto/cs/apluscourses/ui/repl/ReplConfigurationForm.java (hunk 4 above)
Comment (OlliKiljunen) [reply src/main/java/fi/aalto/cs/apluscourses/ui/repl/ReplConfigurationForm.java:47]: The (renamed) method was initially declared static though it shouldn't have been. `PluginSettings` is a singleton class.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java (hunk 5 above)
Comment (OlliKiljunen) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:99]: I've understood "default project" is some kind of placeholder object for "no project open" situation but I don't really know in which circumstances this actually occurs. Something we should research on at some point.

**Comment by superseacat** [comment]
> [](https://sonarcloud.io/component_measures?id=Aalto-LeTech_intellij-plugin&pullRequest=142&metric=new_coverage&view=list) [6.9% Coverage](https://sonarcloud.io/component_measures?id=Aalto-LeTech_intellij-plugin&pullRequest=142&metric=new_coverage&view=list)

this might need some more attention

**Comment by superseacat** [comment]
> [](https://sonarcloud.io/component_measures?id=Aalto-LeTech_intellij-plugin&pullRequest=142&metric=new_coverage&view=list) [30.1% Coverage](https://sonarcloud.io/component_measures?id=Aalto-LeTech_intellij-plugin&pullRequest=142&metric=new_coverage&view=list)  

can this be improved, e.g. by drafting some tests for [SettingsImporterImpl.java](https://github.com/Aalto-LeTech/intellij-plugin/blob/create-course-project/src/main/java/fi/aalto/cs/apluscourses/intellij/model/SettingsImporterImpl.java)?

**Comment by nikke234** [comment]
> > [](https://sonarcloud.io/component_measures?id=Aalto-LeTech_intellij-plugin&pullRequest=142&metric=new_coverage&view=list) [30.1% Coverage](https://sonarcloud.io/component_measures?id=Aalto-LeTech_intellij-plugin&pullRequest=142&metric=new_coverage&view=list)
> 
> can this be improved, e.g. by drafting some tests for [SettingsImporterImpl.java](https://github.com/Aalto-LeTech/intellij-plugin/blob/create-course-project/src/main/java/fi/aalto/cs/apluscourses/intellij/model/SettingsImporterImpl.java)?

Yes, will add more tests :+1:
